src/work/BasicWork.h

```diff
@@ -0,0 +1,153 @@
+ *  it is implementers' responsibility to implement `onRun`, that
+ *  hints the next desired state.
+ */
+
+class BasicWork : public std::enable_shared_from_this<BasicWork>,
+                  private NonMovableOrCopyable
+{
+  public:
+    static size_t const RETRY_NEVER;
+    static size_t const RETRY_ONCE;
+    static size_t const RETRY_A_FEW;
+    static size_t const RETRY_A_LOT;
+    static size_t const RETRY_FOREVER;
+
+    // Publicly exposed state of work:
+    // WORK_RUNNING - work has been created and is currently in progress
+    // (note that this might mean that work is being destructed also).
+    // Implementers return RUNNING when work needs to be scheduled to run more.
+    // WORK_WAITING - work should not be scheduled to run, as it's waiting
+    // for some event (process exit, retry timeout, etc)
+    // Implementers return WAITING when they don't want the work to be
+    // scheduled. Note that it is implementers' responsibility then to correctly
+    // use `wakeUp` to exit WAITING state. WORK_SUCCESS - terminal successful
+    // state WORK_FAILURE - terminal unsuccessful state. When this state is
+    // returned, depending on the Work's retry strategy, either a retry will be
+    // scheduled or work will cease execution.
+    enum class State
+    {
+        WORK_RUNNING,
+        WORK_WAITING,
+        WORK_SUCCESS,
+        WORK_FAILURE,
+    };
+
+    BasicWork(Application& app, std::string name, size_t maxRetries);
+    virtual ~BasicWork();
+
+    std::string const& getName() const;
+    virtual std::string getStatus() const;
+    State getState() const;
+    bool isDone() const;
+
+    size_t getMaxRetries() const;
+    uint64_t getRetryETA() const;
+
+    // Main method for state transition, mostly dictated by `onRun`
+    void crankWork();
+
+    // Reset work to its initial state (only if work is in terminal state)
+    void startWork(std::function<void()> wakeUpParent);
+    virtual void shutdown();
+
+  protected:
+    // Implementers can override these callbacks to customize functionality
+    // `onReset` is expected to restore work's state to its initial condition;
+    // this includes cleaning up side effects (e.g. produced files), restoring
+    // member variables etc.
+    // `onRun` performs necessary work logic, and hints the
+    // next state transition. `onSuccess` is called when work transitions into
+    // WORK_SUCCESS state. Similarly, `onFailure*` methods are called upon
+    // transitioning into appropriate failure state. An example usage would be
+    // cleaning up side effects, like downloaded files in these methods. If you
+    // want to force failure (and reset / retry) you can return
+    // WORK_FAILURE_RETRY or WORK_FAILURE_RAISE. After a retry count is
+    // passed, WORK_FAILURE_RETRY means WORK_FAILURE_RAISE anyways.
+
+    virtual void onReset();
+    virtual State onRun() = 0;
+    virtual void onWakeUp();
+    virtual void onFailureRetry();
```

Comment:
This one seems not to be used.
Probably only onFailureRaise is needed (and can be renamed to onFailure).

Reply:
Looks like it's needed for RepairMissingBucketsWork (as it needs to differentiate between a retry vs terminal failure to call a handler)